### PR TITLE
Fixes #351 Add serializer for passing non-string values from python to bash and vice versa

### LINKED ISSUE #351: Implement a basic string serializer and deserializer for better communication between config.py and conf.py
### This is a
- [ ] Bug report
- [x] Feature request

### Current behaviour

Currently, the configuration read using config.py, is serialized to a string format using basic rules such as joining list via comma, converting None to empty string, etc. And then these values are passed to conf.py from generate.sh using bash variables.

### Expected behaviour

Implement a basic serializer moving all conversion logic "from python objects to strings and vice versa" to a different module.

### PR DESCRIPTION
**Pull request by pri22296**
## Description

- Theme options and extension's configuration can now be non-strings. Earlier, to pass non string theme options, one had to use a manual conf.py
- As a result, the syntax for theme options has also been changed. Now instead of `=` seperated strings they can be regular yaml based key, value pairs. The values can also be yaml constructs like lists and they would be correctly translated to appropriate python objects. Currently supported python objects are `boolean`, `str`, `list`, `None`.

Earlier,

```yaml
theme:
  options:
    - "option1=value1"
```

Now,
```yaml
theme:
  options:
    option1: value1
    option2:
      - value1
      - value2
```
Note that the second option wasn't supported in the earlier implementation.

- As a result, the bash variables encoding lists, would have a surrounding a `[` and `]` around them. scripts would need to handle it properly when using them.

## Related Issue

Fixes #351 

## How Has This Been Tested?

Deployment: https://yaydocmd.herokuapp.com/

## Types of changes

- [ ] Bug fix 
- [x] New feature

## Checklist:

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] There is a corresponding issue for this pull request.
- [x] Mentioned the Issue number in the pull request commit message `Fixes #<number> commit message`
- [x] There is only one commit per issue.

### CODE REVIEW

**Comment by pri22296** [comment]
@sch00lb0y @imujjwal96 Please review

File: backend/generator.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ exports.executeScript = function (socket, formData, callback) {
     "-d", debug,
     "-u", uniqueId,
     "-s", subProject,
-    "-p", subDocpath.join(","),
+    "-p", "[" + subDocpath.join(",") + "]",
```
Comment (imujjwal96) [inline backend/generator.js]: why was this change needed?

Comment (pri22296) [inline backend/generator.js]: Otherwise there is no way to detect whether a single string is a list of values or a string. This is required because of the following yaml syntax.

Let's say a field takes a set of values:
```yaml
field:
     - value1
     - value2
```
When user needs to only pass a single value, they expect that the following syntax would also work.
```yaml
field: value1
```
To handle such situation we need to pass that information encoded in the string passed to bash so that from the conf.py. That is why this basic serializer encloses the comma seperated string with square brackets. Now since that string will also be need to deserialized, we'd need to enclose it ourselves when manually setting value for any variable such as in this case.

Comment (imujjwal96) [inline backend/generator.js]: Thanks a lot for the explanation. It is highly appreciated.

**Comment by imujjwal96** [comment]
How should I test it? Can you show and compare deployments without this change and with it? I will be grateful.

**Comment by pri22296** [comment]
@imujjwal96 This is an internal change so not much will be visible from outside. You can try out the new theme options syntax and I've also added some tests related to the serializer which you can check out.

To test the theme options syntax:
You can test one of the theme options provided by alabaster which are described here. http://alabaster.readthedocs.io/en/latest/customization.html#variables-and-feature-toggles

**Comment by pri22296** [comment]
@sch00lb0y @imujjwal96 Could you review this once more please? I don't know why github marks a review as stale on a rebase.
